### ISSUE TITLE
Modal Dialog should address onClose focus behavior

### ISSUE THREAD

**Issue by nicholasrice** [microsoft/fast-dna#2565]
In investigating #2536 I discovered that the modal dialog does no work to focus an element on close.

As described [here](https://www.w3.org/TR/wai-aria-practices/#dialog_modal), in most cases a Modal Dialog has a well defined focus behavior when the dialog is closed. It doesn't seem to me that our current implementation of Base Dialog aligns to our goal that base components meet accessibility requirements (where reasonable). I think we can do better here by adding a single prop *focusOnClose*

```
focusOnClose: boolean | React.RefObject<HTMLElement> | HTMLElement // default to `true`
```

By default, the dialog would store a reference to the active document element when opening, prior to moving focus to the dialog. When closing, assuming the stored element still exists, the element would be focused. I believe this would address the majority of Modal Dialog cases and matches the default behavior of Modal Dialogs. 

Alternatively, an `HTMLElement` or `React.RefObject` could be provided to the same prop. If that prop is provided, the provided element / refObject would be focused instead. This would help address custom element focusing on dialog close.

A potential flow exists where the element that needs to be focused might not exist until after the dialog close. In this case, the `focusOnClose` prop would be set to `false` (the dialog would not perform any onClose focus behavior). App authors could then perform any necessary work from the `onDismiss` callback that already exists on the Dialog.

Any thoughts or feedback is welcome.

**Comment by chrisdholt** [comment]
Looks good - it would be ideal to expedite this to solve the problems which are out there.

**Comment by scomea** [comment]
"App authors could then perform any necessary work from the onDismiss callback that already exists on the Dialog."  This is not actually what the callback does.  It merely informs the author that the dialog has detected an event (ie. esc key press) and that the author's code might want to dismiss the dialog.  At this point the dialog would still exist and still steal focus if something outside the dialog gets it.  I'm not dead set against the focus mechanism you suggest, but I worry we'd be making a lot of assumptions but I'm thinking a straight up "onDialogUnmounted" callback that lets authors do what they want.

**Comment by nicholasrice** [comment]
> "App authors could then perform any necessary work from the onDismiss callback that already exists on the Dialog." This is not actually what the callback does. 

I'm pretty sure we're on the same page here - my understanding is that the `onDismiss` callback is intended to inform app authors that the user has taken action to dismiss the dialog, so they can make the appropriate upstream changes to *not* render the dialog (through unmounting or the visible prop).

My hesitation for an `onDialogUnmounted` callback is that it is a prop interface directly tied to the implementation details of a React component rather then the semantic requirements for the piece of UI. There are two methods of removing a visible modal dialog - setting `visible={false}` and *not rendering the dialog* (`{this.state.renderDialog ? <Dialog ... /> : null }`) - In an ideal world, I think the both mechanism should be behaviorally identical to an app author, which I don't think a `onDialogUnmount` prop supports, because setting `visible={false}` does not unmount the component.

Can you be more specific about what assumptions you're worried about? I *believe* the outlined approach provides reasonable default behavior, provides multiple mechanisms to override, and supports all the behavioral requirements of the dialog, but we should certainly address any unmet requirements or unintuitive behavior.

**Comment by scomea** [comment]
Yes, I would expect the "dialog inactive" callback to happen not just on dismount but at the point where it would no longer take focus regardless of how that came about.  The assumption I worry about is regarding the default behavior of grabbing an arbitrary element ref and deciding we should force focus back to it later by default.  I worry that this could lead to some hard to diagnose/reproduce bugs for someone who isn't paying attention that the dialog is doing this.  The element may not even still be in the dom when the dialog is dismissed, or the UI may have changed and focusing on it scrolls the ui to bring it into view, for example.

**Comment by nicholasrice** [comment]
The element wouldn't be arbitrary - it would be the element that had focus prior to the dialog opening. In cases where the element in question triggers the Dialog, this is the behavior outlined in by [ARIA](https://www.w3.org/TR/wai-aria-practices/#dialog_modal). If the element is *not* responsible for invoking the dialog (eg Dialog is on a timer) we've *at least* returned the user to their prior location. In cases where the element doesn't exist when the dialog closes - we would certainly ensure it exists before invoking focus. Perhaps we can warn the developer if they *also* have not supplied a focus element?

We could also not have default behavior, I just worry that we wouldn't be going far enough to address the common scenarios and associated accessibility requirements.

**Comment by scomea** [comment]
The timer scenario is one of those I was thinking about, - where focus isn't obviously anchored to the ui that invoked that dialog.

**Comment by chrisdholt** [comment]
I would advocate here for default behavior matching what @nicholasrice is describing. Supporting and aligning to the ARIA specification is foundational to how we build our components, so I don't think it's unreasonable for us to handle this internally. In this case, we can programmatically have a reference to the element which triggered the dialog which provides us what we need to check that it's available and can have focus returned when the dialog is dismissed/closed. This seems reasonable so long as we cover our bases (such as the element existing and the timer scenario on load, etc).

**Comment by scomea** [comment]
Can we be confident that a reference to a dom element we get from grabbing the document active element is still in that role several render cycles later? I'm not totally clear on when React reuses dom nodes or not.

**Comment by scomea** [comment]
After discussing with Nick we came up with a rough consensus that we could have a single prop ("refocusTarget"?) that takes an HTMLElement, Ref<HTMLElement>, or () => void so authors can provide an element to focus on or have a callback to prompt them. We had anticipated making this a required prop (and not have a default focus behavior so we don't get in the business of trying to figure out where focus should go) but in retrospect that would seem odd with non-modal dialogs.

**Comment by nicholasrice** [comment]
Agreed with @scomea. We can add it as an optional prop for the time being, and address how to address the non-modal case when it comes time to make a breaking change. One option would be to have a Modal component vs a non-modal dialog - behaviorally they are different so multiple components could make sense.

**Comment by chrisdholt** [comment]
This was closed with #2603
